### DISCUSSION TITLE
participle in the field!

### DISCUSSION THREAD

**Discussion by drew-512** [alecthomas/participle#173] [Q&A]
Hi friends and @alecthomas, very impressive work!  First allow me to say thanks for gifting it to the world.  I love Go and projects like these remind of me of why.  :)

I'm leading development of an exciting particle physics project that will be presented a number of research institutions including the LHCb in the coming year, and `participle` seems like a great fit to handle queries from users (particle physicists) into our package.  I'm getting the hang of it, but using participle would go way faster if I had help with a starting point for a simple stateless expr language:

```
    Graph:  <Vtx>[<Edge><Vtx>]*[, ...]
    Edge:   "-" | "~" | "=" | "--" | "~~"
    Vtx:    <VtxID>["^" | "^^" | "^^^" ]
    VtxID:  uint8 (> 0)      
```

Basically, this is a language to express a graph, such as:
| Example `string`      | # Edge Runs |
| ----------- | ----------- |
| 1             |  1 |
| 1-2-3      | 1 |
| 1-2-3, 2-4   | 2 |
| 1^-2=3^      | 1 |
| 1-2, 1-3 | 2 |
| 1=2, 1~~3  | 2 | 

Some nice-to-haves:
  - Support for /* */ style inline comments (that are ignored) 
  - Instead of commas delimiting edge "runs", they could instead be optional (and a run would be closed by whitespace or EOS).  In general, we want whitespace to be insertable anywhere to help with readability and to prevent physicists from getting tripped up on whitespace nuances.

Once parsed, we would then walk each Graph, perhaps something like:
```
    for run := range g.EdgeRuns() {
        // walk though `run`, extracting alternating VtxID and edge types
    }
```

We will be unveiling the project in about 2 months and it will be GPL3.  We would love to shoutout to `participle`  and publicly acknowledge it.  This is a passion project and are not getting paid, so any help would be greatly appreciated.

Hugs,
Drew

**Comment by alecthomas** [comment]
I'm glad you're enjoying Participle. Have you been through the [tutorial](https://github.com/alecthomas/participle/blob/master/TUTORIAL.md)? If you do that you should have basically all of the steps necessary to create a parser for this structure, particularly as you already have it in an EBNF-style form.

Basically each of the rules should end up being a struct. The start of the grammar would look something like this:

```go
type Graph struct {
  Spans []*Span `(@@ ("," @@)*)?`
}

type Span struct {
  Start *Vtx `@@`
  Edge []*Edge `@@+`
}

type Edge struct {
  Kind string `@("-" "-"? | "~" "~"? | "="  )`
  End *Vtx `@@`
}

...
```

The default lexer should be sufficient I believe but will return individual characters for punctuation, never runs of characters. The parser only matches exact tokens. So `"-" "-"` will work but `"--"` will not.

**Reply by drew-512** [reply]
Interesting, got it!  Thanks for your support man.  Hm, so looks like my only option to get creative it seems.  There are 10 different types of edge types.  If you are bored and like hearing about this, the project is super exciting and addicting.  You have been warned.  

Anyway, this is me seeing what a unicode grammar would look like vs keepin it all in ASCII baby.  Anything strike you?

```go

// EdgeType names the type of edge
type EdgeType byte
const (
	NilEdge       EdgeType = 0
	PosEdge       EdgeType = (1 << 2)
	NegEdge       EdgeType = (1 << 2) | 1
	PosPosEdge    EdgeType = (2 << 2)
	PosNegEdge    EdgeType = (2 << 2) | 1
	NegNegEdge    EdgeType = (2 << 2) | 2
	PosPosPosEdge EdgeType = (3 << 2)
	PosPosNegEdge EdgeType = (3 << 2) | 1 
	PosNegNegEdge EdgeType = (3 << 2) | 2
	NegNegNegEdge EdgeType = (3 << 2) | 3
	
	EdgeTypeMask EdgeID = 0x000F // low 4 bits
)

func (et EdgeType) Symbol() string {
	return [16]string{
		" ", " ", " ", " ", 
		"-", "~", " ", " ", 
		"=", "-~", "~~", " ", 
		"---", "--~", "-~~", "~~~",
		}[et]
}

func parseEdgeSymbol(str string) (EdgeType, int, error) {
	pos := byte(0)
	neg := byte(0)
	i := 0
	for _, r := range str {
		switch r {
		case '-': pos += 1
		case '~': neg += 1
		
		case '=': pos += 2
		case '≃': pos += 1; neg += 1
		case '≈': neg += 2

		case '≡': pos += 3
		case '≅': neg += 1; pos += 2
		case '≊': neg += 2; pos += 1		
		case '≋': neg += 3
		
		default:
			if i == 0 {
				return NilEdge, 0, nil
			}
			break
		}
		i++
	}
	total := pos + neg
	edgeType := EdgeType((total << 2) | neg)
	return edgeType, i, nil
}

```

**Reply by drew-512** [reply]
...so I was thinking ASCII since I feel like we're doing this project in the name of science and everything.   But then that means, ya know.  It's super important this grammar is something physicists can read and type ad hoc easily.  shrug

**Reply by drew-512** [reply]
Oh, and some of those poor damn chars like `≊` are barely readable in almost any font nor can be easily typed (as in fingers).

**Reply by drew-512** [reply]
p.s. I think my mind is beautifully blown by starting to see the world and parsing and ASTs in a huge new way.  Wild to see things this way.

**Reply by alecthomas** [reply]
> p.s. I think my mind is beautifully blown by starting to see the world and parsing and ASTs in a huge new way. Wild to see things this way.

Yes! That was exactly my reaction when I thought of the Participle approach initially :)

**Reply by drew-512** [reply]
I guess there's no way to get Go to shush on `reflect` warnings without taking out the whole warning?

```
struct field tag `@@` not compatible with reflect.StructTag.Get: bad syntax for struct tag pair
```
